bfabric/bfabric2.py

```diff
@@ -173,71 +217,104 @@ def read(self, endpoint: str, obj: dict, max_results: Optional[int] = 100, readi
         :param check: whether to check for errors in the response
         :return: List of responses, packaged in the results container
         """
+        # Ensure stability
+        obj = self._add_query_timestamp(obj)
 
         # Get the first page.
         # NOTE: According to old interface, this is equivalent to plain=True
-        response = self._read_method(readid, endpoint, obj, page=1, **kwargs)
+        response, errors = self._read_page(readid, endpoint, obj, page=1)
+
         try:
-            n_pages = response["numberofpages"]
+            n_available_pages = response["numberofpages"]
         except AttributeError:
-            n_pages = 0
+            n_available_pages = 0
 
         # Return empty list if nothing found
-        if not n_pages:
-            result = ResultContainer([], self.result_type, total_pages_api=0, errors=get_response_errors(response, endpoint))
+        if not n_available_pages:
+            result = ResultContainer(
+                [], self.result_type, total_pages_api=0, errors=get_response_errors(response, endpoint)
+            )
             if check:
                 result.assert_success()
             return result
 
         # Get results from other pages as well, if need be
-        # Only load as many pages as user has interest in
-        if max_results is None:
-            n_pages_trg = n_pages
-        else:
-            n_pages_trg = min(n_pages, div_int_ceil(max_results, BFABRIC_QUERY_LIMIT))
+        requested_pages, initial_offset = compute_requested_pages(
+            n_page_total=n_available_pages,
+            n_item_per_page=BFABRIC_QUERY_LIMIT,
+            n_item_offset=offset,
+            n_item_return_max=max_results,
+        )
+        logging.info(f"Requested pages: {requested_pages}")
 
         # NOTE: Page numbering starts at 1
-        response_items = response[endpoint]
-        errors = []
-        for i_page in range(2, n_pages_trg + 1):
-            print('-- reading page', i_page, 'of', n_pages)
-            response = self._read_method(readid, endpoint, obj, page=i_page, **kwargs)
-            errors += get_response_errors(response, endpoint)
-            response_items += response[endpoint]
-
-        result = ResultContainer(response_items, self.result_type, total_pages_api=n_pages, errors=errors)
+        response_items = []
+        page_offset = initial_offset
+        for i_iter, i_page in enumerate(requested_pages):
+            if not (i_iter == 0 and i_page == 1):
```

Comment:
Corner case: i_iter == 0 and i_page != 1, `errors` could be misleading...

Reply:
I guess it could be argued, that it's best to return all errors that were encountered in the processing of the method (and it partially feels like something that has low practical relevance).